subxt/src/constants/constant_address.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use crate::{
+    dynamic::DecodedValue,
+    metadata::DecodeWithMetadata,
+};
+use std::borrow::Cow;
+
+/// This represents a constant address. Anything implementing this trait
+/// can be used to fetch constants.
+pub trait ConstantAddress {
+    /// The target type of the value that lives at this address.
+    type Target: DecodeWithMetadata;
+
+    /// The name of the pallet that the constant lives under.
+    fn pallet_name(&self) -> &str;
+
+    /// The name of the constant in a given pallet.
+    fn constant_name(&self) -> &str;
+
+    /// An optional hash which, if present, will be checked against
+    /// the node metadata to confirm that the return type matches what
+    /// we are expecting.
+    fn validation_hash(&self) -> Option<[u8; 32]> {
+        None
```

Comment:
why default impl? 

I have been bitten by that a couple of times and prefer explict impls for each type :P

Reply:
This is just because we only really ever expect to use `validation_hash` internally to allow us to check our static codegen against our metadata hashes. By default, anybody who implements this trait themselves will not be doing their own static codegen and will instead be using their own custom types which won't have a sensible impl for this :)

The worst case of not implementing it is that we don't do our extra validation check anyway